File: NetNodePluginClient/blue.py
import threading

from EtherealC.Client.WebSocket.WebSocketClient import WebSocketClient
from EtherealC.Core.Model.AbstractTypes import AbstractTypes
from EtherealC.Core.Model.TrackException import TrackException, ExceptionCode
from EtherealC.Net.Abstract.Net import Net
from EtherealC.Net.WebSocket.WebSocketNetConfig import WebSocketNetConfig


class WebSocketNet(Net):
    def __init__(self, name):
        super().__init__(name=name)
        self.config = WebSocketNetConfig()
        self.connectSign = threading.Event()

    def Publish(self):
        def reactorStart():
            from twisted.internet import reactor
            if not reactor.running:
                reactor.suggestThreadPoolSize(10)
                reactor.run(False)

        threading.Thread(target=reactorStart).start()
        if self.config.netNodeMode is True:
            types = AbstractTypes()
            types.add(type=int, type_name="Int")
            from EtherealC.Net.NetNodeClient.Model.NetNode import NetNode
            types.add(type=type(NetNode()), type_name="NetNode")
            types.add(type=str, type_name="String")
            types.add(type=bool, type_name="Bool")
            from EtherealC.Service import ServiceCore
            from EtherealC.Net.NetNodeClient.Service.ClientNetNodeService import ClientNetNodeService
            netNodeService = ServiceCore.Register(service=ClientNetNodeService(name="ClientNetNodeService", types=types), net=self)
            from EtherealC.Request import RequestCore
            from EtherealC.Net.NetNodeClient.Request.ServerNetNodeRequest import ServerNetNodeRequest
            netNodeRequest = RequestCore.Register(net=self, request=ServerNetNodeRequest(name="ServerNetNodeService", types=types))

            def NetNodeSearchRunner():
                try:
                    from EtherealC.Net import NetCore
                    while NetCore.Get(self.name) is not None:
                        self.NetNodeSearch()
                        self.connectSign.wait(timeout=self.config.netNodeHeartbeatCycle/1000)
                        self.connectSign.clear()
                except Exception as exception:
                    self.OnException(
                            TrackException(exception=exception, message="NetNodeSearch循环报错", code=ExceptionCode.Runtime))

            threading.Thread(target=NetNodeSearchRunner).start()
        else:
            try:
                for request in self.requests.values():
                    request.client.Connect()
            except Exception as e:
                self.OnException(exception=TrackException(exception=e))
        return True

    def NetNodeSearch(self):
        flag = False
        for request in self.requests.values():
            if request.client is None and request.name != "ServerNetNodeService":
                flag = True
                break
        if flag is True:
            for item in self.config.netNodeIps:
                prefixes = item["prefixes"]
                config = item["config"]
                from EtherealC.Client import ClientCore
                client = ClientCore.Register(net=self, service_name="ServerNetNodeService",client=WebSocketClient(prefixes=prefixes))
                client.Connect(isSync=True)
                try:
                    if client.IsConnect():
                        from EtherealC.Request import RequestCore
                        from EtherealC.Net.NetNodeClient.Request.ServerNetNodeRequest import ServerNetNodeRequest
                        netNodeRequest: ServerNetNodeRequest = RequestCore.Get(net=self,
                                                                               service_name="ServerNetNodeService")
                        if netNodeRequest is None:
                            raise TrackException(code=ExceptionCode.Runtime, message="无法找到{0}-ServerNetNodeService"
                                                 .format(self.name))
                        for request in self.requests.values():
                            if request.client is not None:
                                continue
                            node = netNodeRequest.GetNetNode(request.name)
                            if node is None:
                                raise TrackException(code=ExceptionCode.Runtime,
                                                     message="{0}-{1}-在NetNode分布式中未找到节点"
                                                     .format(self.name, request.name))
                            requestClient = ClientCore.Register(net=self, service_name=request.name,
                                                                client=WebSocketClient(prefixes=prefixes))
                            requestClient.disconnect_event.register(self.ClientConnectFailEvent)
                            requestClient.Connect()
                        break
                finally:
                    ClientCore.UnRegister(net=self, service_name="ServerNetNodeService")

    def ClientConnectFailEvent(self, client):
        from EtherealC.Client import ClientCore
        ClientCore.UnRegister(net=self, service_name=client.name)
        self.connectSign.set()

File: NetNodePluginClient/blue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NetNodePluginClient
{
    public class blue
    {
        /*
        void b()
        {
            //注册数据类型
            AbstractTypes types = new AbstractTypes();
            types.Add<int>("Int");
            types.Add<long>("Long");
            types.Add<string>("String");
            types.Add<bool>("Bool");
            types.Add<NetNode>("NetNode");
            //向网关注册服务
            Service.Abstract.Service netNodeService = ServiceCore.Register(this, new ClientNetNodeService(),
                "ClientNetNodeService", types);
            //向网关注册请求
            ServerNetNodeRequest netNodeRequest =
                RequestCore.Register<ServerNetNodeRequest, IServerNetNodeRequest>(this, "ServerNetNodeService",
                    types);
            new Thread(() =>
            {
                try
                {
                    while (NetCore.Get(name, out Abstract.Net net))
                    {
                        NetNodeSearch();
                        connectSign.WaitOne(Config.NetNodeHeartInterval);
                    }
                }
                catch (Exception e)
                {
                    OnException(new TrackException(e));
                }
            }).Start();
        }
        /// <summary>
        /// 分布式模式Demo
        /// </summary>
        /// <param name="netName">网关名</param>
        /// <param name="ip">本地集群地址</param>
        public static void NetNode(string netName,string ip)
        {
            //注册数据类型
            AbstractTypes types = new AbstractTypes();
            types.Add<int>("Int");
            types.Add<User>("User");
            types.Add<long>("Long");
            types.Add<string>("String");
            types.Add<bool>("Bool");
            //建立网关
            Net net = NetCore.Register(new WebSocketNet(netName));
            net.ExceptionEvent += Config_ExceptionEvent;
            //向网关注册服务
            Service service = ServiceCore.Register(net, new ClientService(), "Client", types);
            //向网关注册请求
            Request request = RequestCore.Register<ServerRequest,IServerRequest>(net, "Server", types);
        //开启分布式模式
        net.Config.NetNodeMode = true;
            //添加分布式地址
            List<Tuple<string, ClientConfig>> ips = new();
        ips.Add(new Tuple<string, ClientConfig>($"ethereal://{ip}:{28015}/NetDemo/", new WebSocketClientConfig()));
            ips.Add(new Tuple<string, ClientConfig>($"ethereal://{ip}:{28016}/NetDemo/", new WebSocketClientConfig()));
            ips.Add(new Tuple<string, ClientConfig>($"ethereal://{ip}:{28017}/NetDemo/", new WebSocketClientConfig()));
            ips.Add(new Tuple<string, ClientConfig>($"ethereal://{ip}:{28018}/NetDemo/", new WebSocketClientConfig()));
            net.Config.NetNodeIps = ips;
            request.ConnectSuccessEvent += Request_ConnectSuccessEvent;
            net.Publish();
            Console.Read();
        }
        public async void NetNodeSearch()
        {
            bool flag = false;
            foreach (Request.Abstract.Request request in Requests.Values)
            {
                if (request.Client == null && request.Name != "ServerNetNodeService")
                {
                    flag = true;
                    break;
                }
            }
            if (flag)
            {
                //搜寻正常启动的注册中心
                foreach (Tuple<string, ClientConfig> item in Config.NetNodeIps)
                {
                    string prefixes = item.Item1;
                    ClientConfig config = item.Item2;
                    //向网关注册连接
                    WebSocketClient client = (WebSocketClient)ClientCore.Register(this, "ServerNetNodeService", new WebSocketClient(prefixes));
                    try
                    {
                        await client.ConnectSync();
                        //连接成功
                        if (client?.Accept?.State == System.Net.WebSockets.WebSocketState.Open)
                        {
                            if (RequestCore.Get(this, "ServerNetNodeService",
                                out Request.Abstract.Request netNodeRequest))
                            {
                                foreach (Request.Abstract.Request request in Requests.Values)
                                {
                                    if (request.Client == null)
                                    {
                                        //获取服务节点
                                        NetNode node = (netNodeRequest as ServerNetNodeRequest).GetNetNode(request.Name);
                                        if (node != null)
                                        {
                                            //注册连接并启动连接
                                            Client.Abstract.Client requestClient = ClientCore.Register(request, new WebSocketClient(node.Prefixes[0]));
                                            requestClient.ConnectFailEvent += ClientConnectFailEvent;
                                            requestClient.DisConnectEvent += ClientDisConnectEvent; ;
                                            requestClient.Connect();
                                        }
                                        else
                                            throw new TrackException(TrackException.ErrorCode.Runtime,
                                                $"{name}-{request.Name}-在NetNode分布式中未找到节点");
                                    }
                                }
                                return;
                            }
                            throw new TrackException(TrackException.ErrorCode.Runtime, $"无法找到{name}-ServerNetNodeService");
                        }
                    }
                    finally
                    {
                        ClientCore.UnRegister(this, "ServerNetNodeService");
                    }
                }
            }
        }

        private void ClientDisConnectEvent(Client.Abstract.Client client)
        {
            ClientCore.UnRegister(client.NetName, client.ServiceName);
            connectSign.Set();
        }

        private void ClientConnectFailEvent(Client.Abstract.Client client)
        {
            ClientCore.UnRegister(client.NetName, client.ServiceName);
        }
        */
    }
}
